From 346c1bb5e2aec4fdee191221f3851aa09be8cc35 Mon Sep 17 00:00:00 2001
From: Dmitry Petrochenko <dmitry.petrochenko@intel.com>
Date: Mon, 3 Jun 2013 12:13:02 +0700
Subject: Dalvik: Fix stack overflow check in JIT (when interpStackStart>0x80000000)

BZ: 112701

The stack overflow check had a signed comparison for fp and interpStackEnd,
causing occurence of false stack overflow event when interpStackStart is greater
then 0x80000000 (sign bit enabled) due to signed comparison instruction.
This patch utilizes unsigned comparison in JIT's stack overflow check.

Category: device-enablement
Domain: AOSP-Dalvik-Compiler-CG
Origin: internal
Upstream-Candidate: no, needs rework

Change-Id: I92e998571ed0f5a6ea1819394dde17ae997acd92
Orig-MCG-Change-Id: I3ddd193781d51ef8009ff512bbb6cd70b76cde4e
Signed-off-by: Dmitry Petrochenko <dmitry.petrochenko@intel.com>
Signed-off-by: Qiming Shi <qiming.shi@intel.com>
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 vm/compiler/codegen/x86/LowerInvoke.cpp |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/vm/compiler/codegen/x86/LowerInvoke.cpp b/vm/compiler/codegen/x86/LowerInvoke.cpp
index d294c86..b51d34a 100644
--- a/vm/compiler/codegen/x86/LowerInvoke.cpp
+++ b/vm/compiler/codegen/x86/LowerInvoke.cpp
@@ -1000,7 +1000,7 @@ int common_invokeArgsDone(ArgsDoneType form) {
             PhysicalReg_EAX, true, PhysicalReg_ESI, true);
 
     // Handle frame overflow
-    conditional_jump(Condition_L, ".stackOverflow", true);
+    conditional_jump(Condition_B, ".stackOverflow", true);
 
     if (form == ArgsDone_Full) {
         // Check for a native call
-- 
1.7.4.1

